src/fs/syslib/kfile.c

```diff
@@ -47,12 +47,13 @@ struct file_desc *kopen(struct node *node, int flag) {
 		return NULL;
 	}
 
-	if (node_is_file(node)) {
+	if (!(node_is_directory(node)) && !(node_is_file(node))) {
 		if (NULL == nas->fs->drv) {
 			SET_ERRNO(ENOSUPP);
 			return NULL;
 		}
-		ops = (struct kfile_operations *) nas->fs->drv->file_op;
+		extern struct idesc *char_dev_open(struct node *node, int flags);
+		return char_dev_open(node, flag);
```

Comment:
I'm not sure but shouldn't we create here `file_desc` firstly and only then return idesc? I think so because in functions below (`kwrite` etc.) `file_desc` will be passed as an argument.

Reply:
Probably it's ok, since we can guarantee `kwrite` is only used for usual files, right?. But it's not obvious because of the similar naming of `kopen`/`kwrite`. Assuming I called `kopen` I'm hoping to see `kwrite` somewhere.

And direct calling of `char_dev_open` here looks a bit weird..

Reply:
`idesc` used instead of `file_desc` need some consideration.
Direct call of `char_dev_open` is needed because there's no other way to interact with `char_dev`